jax/_src/nn/functions.py

```diff
@@ -765,3 +768,203 @@ def hard_silu(x: ArrayLike) -> Array:
   return x_arr * hard_sigmoid(x_arr)
 
 hard_swish = hard_silu
+
+def _causal_mask(T, dtype):
+  dtype_max = jnp.finfo(dtype).max
+  large_negative_number = jnp.asarray(-0.5 * dtype_max, dtype=dtype)
+  col_idx = jnp.tile(jnp.arange(T)[jnp.newaxis, :], [T, 1])
+  row_idx = jnp.tile(jnp.arange(T)[:, jnp.newaxis], [1, T])
+  mask = (row_idx < col_idx).astype(dtype) * large_negative_number
+  mask = mask[jnp.newaxis, jnp.newaxis, :, :]
+  return mask
+
+class SdpaCausalMask:
+  def __call__(self, x: ArrayLike, mask: ArrayLike | None = None) -> Array:
+    """Mask the input with causal masking"""
+    if mask is not None:
+      warnings.warn(
+          "The provided mask will be ignored since the causal mask is used")
+      del mask
+    T = x.shape[-2]
+    assert T == x.shape[-1], f'T should equal to S: {T} vs {x.shape[-1]}'
+    assert jnp.issubdtype(x.dtype, jnp.floating), x.dtype
+    mask = _causal_mask(T, x.dtype)
+    return x + mask
+
+@dataclass
+class SdpaPhiloxDropout:
+  rate: float = 0.0
+  seed: int = 123
```

Comment:
This obviously doesn't compose well with the JAX RNG system as it doesn't take a JAX random seed. Could it?

Reply:
I meant: JAX has an API for splitting PRNG keys (you don't work with raw integer seeds directly except seeding the very first `jnp.PRNGKey`). All JAX random functions take either an int or a `jnp.PRNG` object), so this should also I think.